activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -17,22 +17,36 @@ def to_key
 
       # Returns the primary key column's value. If the primary key is composite,
       # returns an array of the primary key column values.
-      def id
-        return _read_attribute(@primary_key) unless @primary_key.is_a?(Array)
-
-        @primary_key.map { |pk| _read_attribute(pk) }
+      if RUBY_VERSION < "3"
+        def id
+          if @_composite_primary_key ||= false
+            @primary_key.map { |pk| _read_attribute(pk) }
+          else
+            _read_attribute(@primary_key)
+          end
+        end
+      else
+        def id
+          if @_composite_primary_key
```

Comment:
The problem is that it would more more method calls, so would end up slower than directly checking `is_a?(Array)` inside `id`.

We're really in micro-optimization territory here, if I didn't have production profile showing this is a big hotspot I wouldn't have bat an eye at the `is_a?(Array)` in the first place.

Reply:
Lets proceed with an instance variable then but I'm curious to see if we can figure out a module-based approach which will result in just different implementations of the method for different models

Reply:
The module would be easy, it's really just a dynamic `include`, the problem is you can't really deactivate it if a subclas switch back to non-composite key :/

Not sure if it's really a legitimate use case though.

We could also make the dynamic module handle both cases 🤔 , this way you only pay the overhead if one of the ancestor has composite keys.

Reply:
>you can't really deactivate it if a subclas switch back to non-composite key :/

We can't deactivate but we can still include the proper module into the `Child` class right?
So the ancestry tree will be something like

```ruby
Child.ancestors
#=> [Child, RegularId, Parent, CompositeId, ...]
```
I could be missing something from a bigger picture